File: 345. Reverse Vowels of a String/_345.py
class Solution:
    def reverseVowels(self, s: str) -> str:
        vowels = ['a','e','i','o','u','A','E','I','O','U']
        myvowels = list()
        for i in s:
            if i in vowels:
                myvowels.append(i)
        
        myvowels.reverse()
        flag = 0
        s = list(s)
        for i,v in enumerate(s):
            if v in vowels:
                s[i] = myvowels[flag]
                flag += 1
        
        return ''.join(s)

    def reverseVowels1(self, s: str) -> str:
        vowels = ['a','e','i','o','u','A','E','I','O','U']
        res = list(s)
        i1 = 0
        i2 = len(res)-1
        while i1<i2:
            if (res[i1] in vowels) and (res[i2] in vowels):
                res[i1], res[i2] = res[i2], res[i1]
                i1+=1
                i2-=1
                continue
            if not (res[i1] in vowels):
                i1+=1
            if not (res[i2] in vowels):
                i2-=1
        return ''.join(res)

    def reverseVowels2(self, s):
        vowels = {'a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'}
        L = list(s)
        i = 0
        j = len(L) - 1
        while i < j:
            while i < j and L[i] not in vowels:
                i += 1
            while j > i and L[j] not in vowels:
                j -= 1
            L[i], L[j] = L[j], L[i] 
            i += 1
            j -= 1
        return ''.join(L)

a = Solution()
print(a.reverseVowels1('hello'))

File: 345. Reverse Vowels of a String/345.go
package leetcode

import(
	"strings"
)

func reverseVowels(s string) string {
	myvowels := make([]rune, 0)
	for _,v := range s{
		if v == 'a' || v == 'e' || v == 'i' || v == 'o' || v == 'u' || v == 'A' || v == 'E' || v == 'I' || v == 'O' || v == 'U'{
			myvowels = append(myvowels,v)
		}
	}
	res := ""
	flag := len(myvowels)-1
	for _,v := range s{
		if v != 'a' && v != 'e' && v != 'i' && v != 'o' && v != 'u' && v != 'A' && v != 'E' && v != 'I' && v != 'O' && v != 'U'{
			res += string(v)
		}else{
			res += string(myvowels[flag])
			flag --
		}
	}
	return res
}

func reverseVowels1(s string) string {

	sByte := []byte(s)
	i, j := 0, len(s)-1
	for i < j {
		for i < j && !isVowel(sByte[i]) {
			i++
		}
		for i < j && !isVowel(sByte[j]) {
			j--
		}

		sByte[i], sByte[j] = sByte[j], sByte[i]
		i++
		j--

	}
	return string(sByte)
}

func isVowel(b byte) bool {
	if b == byte('a') || b == byte('e') || b == byte('i') || b == byte('o') || b == byte('u') ||
		b == byte('A') || b == byte('E') || b == byte('I') || b == byte('O') || b == byte('U') {
		return true
	}
	return false
}

func reverseVowels2(s string) string {
	vowels := "aeiouAEIOU"
	start := 0
	end := len(s) - 1
	st := []rune(s)

	for start < end {
		if strings.ContainsRune(vowels, st[start]) && strings.ContainsRune(vowels, st[end]){
			tmp := st[start]
			st[start] = st[end]
			st[end] = tmp
			start++
			end--
		} else if strings.ContainsRune(vowels, st[start]) {
			end--
		} else if strings.ContainsRune(vowels, st[end]) {
			start++
		} else {
			start++
			end--
		}
	}

	return string(st)
}